### PR TITLE
Bump k8s to v1.19.0-rc.4. Switch to knative forks.

### LINKED ISSUE #713: upgrade client-go to 0.18.x

### PR DESCRIPTION
**Pull request by mflendrich**
**What this PR does / why we need it**:

- Updates Kubernetes depdendencies from v0.17.x to v0.19.0-rc.4 with the following consequences:
    - includes `networking.k8s.io/v1` Ingress (needed by #788)
    - changes multiple APIs by adding `context.Context` arguments (this PR updates relevant call sites and performs context plumbing where it appeared reasonable)
    - breaks our Knative deps because of the API changes
- Replaces our dependency on `knative.dev/serving` with `knative.dev/networking` (which is desirable because it's a smaller library)
- Replaces our knative dependencies with k8s-v1.18-ready forks (to be switched back when upstream Knative starts building with k8s v1.18 libs)
- Fixes a regression caused by v1.19: before this PR, the conversion mechanism (from `extensions/v1beta1` Ingress to `networking.k8s.io/v1beta1` Ingress) relied on the API scheme performing a fallback (recursive field-wise) conversion. This mode of conversion stops working in k8s 1.19. Per 1.19 release notes:
    
    > K8s.io/apimachinery - scheme.Convert() now uses only explicitly registered conversions - default reflection based conversion is no longer available. +k8s:conversion-gen tags can be used with the k8s.io/code-generator component to generate conversions.
    
    These types (`extensions/v1beta1` and `networking.k8s.io/v1beta1` Ingress) are field-wise compatible. This PR introduces an explicit field-wise conversion using JSON as an intermediary, [as advised on k8s sig-api-machinery slack](https://kubernetes.slack.com/archives/C0EG7JC6T/p1598018172011800).

**Which issue this PR fixes**:
- unblocks #788 which requires k8s API update to v0.19
- closes #713 - this PR actually lobs over v0.18 by upgrading to a pre-GA v0.19

**Special notes for your reviewer**:
- Please consider reviewing commit by commit, and merging without squashing - because these two commits solve different problems.
- Pay particular attention to the sites where contexts are not plumbed through, but rely on `context.Background()`

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,37 +3,37 @@ module github.com/kong/kubernetes-ingress-controller
...
-	knative.dev/serving v0.12.1
+	knative.dev/networking v0.0.0-20200818151513-77f604177dc7
+	knative.dev/pkg v0.0.0-20200818021113-c451e3c37f5f
+)
+
+replace (
+	k8s.io/client-go => k8s.io/client-go v0.19.0-rc.4
+
+	knative.dev/networking => github.com/mflendrich/networking v0.0.0-20200820130121-b71940c64af7
+	knative.dev/pkg => github.com/mflendrich/pkg v0.0.0-20200820125340-3493cae5c655
```
Comment (hbagdi) [inline go.mod:38]: Please fork these repos under kong's account and let's use them.

File: go.mod [hunk 2]
```diff
@@ -3,37 +3,37 @@ module github.com/kong/kubernetes-ingress-controller
...
-	github.com/prometheus/client_golang v1.0.0
-	github.com/sirupsen/logrus v1.4.2
+	github.com/mitchellh/mapstructure v1.3.1
+	github.com/prometheus/client_golang v1.7.1
+	github.com/sirupsen/logrus v1.6.0
 	github.com/spf13/pflag v1.0.5
-	github.com/spf13/viper v1.6.2
-	github.com/stretchr/testify v1.4.0
+	github.com/spf13/viper v1.7.0
+	github.com/stretchr/testify v1.5.1
```
Comment (hbagdi) [inline go.mod:20]: Please don't update unrelated deps with this change.

File: internal/ingress/store/store_test.go [hunk 3]
```diff
@@ -49,6 +50,10 @@ func Test_networkingIngressV1Beta1(t *testing.T) {
 			name: "correctly transformers from extensions to networking group",
 			args: args{
 				obj: &extensions.Ingress{
+					TypeMeta: metav1.TypeMeta{
+						APIVersion: "extensions/v1beta1",
+						Kind:       "Ingress",
+					},
 					Spec: extensions.IngressSpec{
```
Comment (hbagdi) [inline internal/ingress/store/store_test.go:57]: Since we are resorting to JSON, can you please add in a test for `Status` as well?

**Comment by hbagdi** [comment]
Also, please rename the commit prefix to `fix` and `chore`.
We loosely follow https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#type.

File: go.mod (hunk 1 above)
Comment (mflendrich) [reply go.mod]: Done

File: internal/ingress/store/store_test.go (hunk 3 above)
Comment (mflendrich) [reply internal/ingress/store/store_test.go:57]: Done

File: go.mod (hunk 2 above)
Comment (mflendrich) [reply go.mod:20]: Those have been pulled in by the updated knative (the fork is based off master to mimic the next release as closely as possible). Should I force them to stay at original versions?

**Comment by mflendrich** [comment]
Thanks for prompt reviews!

All comments (but one from @hbagdi - requesting decision) resolved by applying the suggestion. PTAL now.

**Comment by hbagdi** [comment]
1. Please squash the third commit into the second.
2. Please run `go mod tidy`.

**Comment by mflendrich** [comment]
>     1. Please squash the third commit into the second.
> 
>     2. Please run `go mod tidy`.

Both done.

File: go.mod (hunk 2 above)
Comment (hbagdi) [reply go.mod:20]: No, that's fine then. This is the part of go.mod I don't like.

**Pull request merged**